crates/gpui/src/platform/windows/direct_write.rs

```diff
@@ -0,0 +1,1286 @@
+        ))
+    }
+}
+
+struct StringIndexConverter<'a> {
+    text: &'a str,
+    utf8_ix: usize,
+    utf16_ix: usize,
+}
+
+impl<'a> StringIndexConverter<'a> {
+    fn new(text: &'a str) -> Self {
+        Self {
+            text,
+            utf8_ix: 0,
+            utf16_ix: 0,
+        }
+    }
+
+    fn advance_to_utf8_ix(&mut self, utf8_target: usize) {
+        for (ix, c) in self.text[self.utf8_ix..].char_indices() {
+            if self.utf8_ix + ix >= utf8_target {
+                self.utf8_ix += ix;
+                return;
+            }
+            self.utf16_ix += c.len_utf16();
+        }
+        self.utf8_ix = self.text.len();
+    }
+
+    fn advance_to_utf16_ix(&mut self, utf16_target: usize) {
+        for (ix, c) in self.text[self.utf8_ix..].char_indices() {
+            if self.utf16_ix >= utf16_target {
+                self.utf8_ix += ix;
+                return;
+            }
+            self.utf16_ix += c.len_utf16();
+        }
+        self.utf8_ix = self.text.len();
+    }
+}
+
+impl Into<DWRITE_FONT_STYLE> for FontStyle {
+    fn into(self) -> DWRITE_FONT_STYLE {
+        match self {
+            FontStyle::Normal => DWRITE_FONT_STYLE_NORMAL,
+            FontStyle::Italic => DWRITE_FONT_STYLE_ITALIC,
+            FontStyle::Oblique => DWRITE_FONT_STYLE_OBLIQUE,
+        }
+    }
+}
+
+impl From<DWRITE_FONT_STYLE> for FontStyle {
+    fn from(value: DWRITE_FONT_STYLE) -> Self {
+        match value.0 {
+            0 => FontStyle::Normal,
+            1 => FontStyle::Oblique,
+            2 => FontStyle::Italic,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl Into<DWRITE_FONT_WEIGHT> for FontWeight {
+    fn into(self) -> DWRITE_FONT_WEIGHT {
+        DWRITE_FONT_WEIGHT(self.0 as i32)
+    }
+}
+
+impl From<DWRITE_FONT_WEIGHT> for FontWeight {
+    fn from(value: DWRITE_FONT_WEIGHT) -> Self {
+        FontWeight(value.0 as f32)
+    }
+}
+
+fn get_font_names_from_collection(
+    collection: &IDWriteFontCollection1,
+    locale: &str,
+) -> Vec<String> {
+    unsafe {
+        let mut result = Vec::new();
+        let family_count = collection.GetFontFamilyCount();
+        for index in 0..family_count {
+            let Some(font_family) = collection.GetFontFamily(index).log_err() else {
+                continue;
+            };
+            let Some(localized_family_name) = font_family.GetFamilyNames().log_err() else {
+                continue;
+            };
+            let Some(family_name) = get_name(localized_family_name, locale) else {
+                continue;
+            };
+            result.push(family_name);
+        }
+
+        result
+    }
+}
+
+unsafe fn get_postscript_name_and_font(
+    font_face: &IDWriteFontFace,
+    locale: &str,
+) -> Option<(String, Font, bool)> {
+    let font_face_pointer = font_face as *const IDWriteFontFace;
+    let font_face_3_pointer: *const IDWriteFontFace3 = std::mem::transmute(font_face_pointer);
```

Comment:
Same for here, let's try to construct this in a normal way, and we can't then we should explain why this is valid.

Reply:
I replaced all the `transmute()` with the corresponding `cast()` function (except where `ManuallyDrop` is involved). While `cast()` might slightly impact performance, it provides a more intuitive cpp-like approach.